### ISSUE TITLE
Docker: Compile Suite and Examples

### ISSUE THREAD

**Issue by ax3l** [ComputationalRadiationPhysics/picongpu#829]
[Docker](https://en.wikipedia.org/wiki/Docker_%28software%29) is an easy, open-source way to share complete images of operating systems including software.

The extremely [easy accessebility](https://www.docker.com/tryit) might not only be very useful for our compile suite (multiple os, multiple images with "oldest"/"newest" deps, very easy command line control and access, chroot) but might also be useful (needs evaluation) for a first try without needing to install all dependencies.

Docker containers can also be stacked/depend on each other, e.g., offical docker images for common OS exist.

**Comment by bussmann** [comment]
I would LOVE to see that!

**Comment by ax3l** [comment]
Since docker is using LXC on linux, CUDA should be no problem at all.

Would be interesting if running a linux instance with a CUDA application can be run [on a mac](https://docs.docker.com/installation/mac/), too. And yes, there might be [windows people](https://docs.docker.com/installation/windows/) too - but the VM-ization there might be a problem for CUDA.

Site note: one can spin up docker instances on Amazone AWS easily, so that issue is related to #764.

**Comment by ax3l** [comment]
[rkt app container](https://github.com/coreos/rkt/) (de: [heise](http://www.heise.de/open/meldung/Anwendungs-Container-mit-Rocket-Docker-Konkurrent-stoesst-auf-breite-Zustimmung-2633392.html))

**Comment by ax3l** [comment]
[new blog article](https://devblogs.nvidia.com/parallelforall/nvidia-docker-gpu-server-application-deployment-made-easy) about containerization of cuda applications :) (I want us to be listed [here](https://github.com/NVIDIA/nvidia-docker/wiki/Third-party) :D )

**Comment by ax3l** [comment]
Here is a first `Dockerfile` based on [`nvidia/cuda:7.0-devel`](https://hub.docker.com/r/nvidia/cuda/) building one/all PIConGPU examples in the `release-0.2.0` branch with C++98:
```bash
FROM       nvidia/cuda:7.0-devel
MAINTAINER Axel Huebl <a.huebl[ät]hzdr.de>

ENV        HOME /root
ENV        DEBIAN_FRONTEND noninteractive

# install dependencies provided by packages
RUN        apt-get update \
           && apt-get install -y --no-install-recommends \
              autoconf \
              git \
              pkg-config \
              make \
              libhdf5-openmpi-dev \
              libopenmpi-dev \
              libpng-dev \
              wget \
              zlib1g-dev

# CMake 3.1+
ENV CMAKE_DOWNLOAD_SUM d251b95829d29e4529dfdbc1e676cd0ef86e9544976603db2f5c26364ad71146
RUN        wget -nv --no-check-certificate \
              https://cmake.org/files/v3.1/cmake-3.1.3-Linux-x86_64.tar.gz \
           && echo "$CMAKE_DOWNLOAD_SUM cmake-3.1.3-Linux-x86_64.tar.gz" | sha256sum -c --strict - \
           && tar -xzf cmake-3.1.3-Linux-x86_64.tar.gz -C /usr/bin       --strip-components=2 cmake-3.1.3-Linux-x86_64/bin \
           && tar -xzf cmake-3.1.3-Linux-x86_64.tar.gz -C /usr/share     --strip-components=2 cmake-3.1.3-Linux-x86_64/share \
           && tar -xzf cmake-3.1.3-Linux-x86_64.tar.gz -C /usr/share/doc --strip-components=2 cmake-3.1.3-Linux-x86_64/doc \
           && tar -xzf cmake-3.1.3-Linux-x86_64.tar.gz -C /usr/share/man --strip-components=2 cmake-3.1.3-Linux-x86_64/man \
           && rm cmake-3.1.3-Linux-x86_64.tar.gz

# boost
RUN        wget -nv http://downloads.sourceforge.net/project/boost/boost/1.57.0/boost_1_57_0.tar.gz \
           && tar -xzf boost_1_57_0.tar.gz && rm boost_1_57_0.tar.gz
RUN        cd boost_1_57_0 \
           && ./bootstrap.sh --with-libraries=filesystem,program_options,regex,system,thread,math --prefix=/usr \
           && ./b2 -j4 && ./b2 install
RUN        rm -rf boost_1_57_0

RUN        mkdir build

# pngwriter
RUN        git clone --depth 50 --branch master \
              https://github.com/pngwriter/pngwriter.git
RUN        cd build \
           && cmake -DCMAKE_INSTALL_PREFIX=/usr ../pngwriter \
           && make install
RUN        rm -rf build/*

# libSplash
RUN        git clone --depth 50 --branch master \
              https://github.com/ComputationalRadiationPhysics/libSplash.git
RUN        cd build \
           && cmake -DCMAKE_INSTALL_PREFIX=/usr ../libSplash \
           && make install
RUN        rm -rf build/*

# ADIOS
RUN        wget -nv http://users.nccs.gov/~pnorbert/adios-1.10.0.tar.gz \
           && tar -xzf adios-1.10.0.tar.gz && rm adios-1.10.0.tar.gz
RUN        cd adios-1.10.0 \
           && CFLAGS="-fPIC" ./configure --enable-static --enable-shared \
              --prefix=/usr --with-mpi --with-zlib=/usr --disable-fortran \
           && make && make install
RUN        rm -rf adios-1.10.0

# get PIConGPU sources
RUN        git clone --depth 50 --branch release-0.2.0 \
               https://github.com/ComputationalRadiationPhysics/picongpu.git

# build PIConGPU examples
#RUN        ./picongpu/compile -j 2 -q -l ./picongpu/examples ./build
#RUN        ./picongpu/compile ./picongpu/examples/ThermalTest ./build

RUN        ./picongpu/createParameterSet ./picongpu/examples/ThermalTest ./case1
RUN        cd build && ../picongpu/configure \
           ../case1 && make VERBOSE=true install
```

Next post: how to rent an AWS instance, install [nvidia-docker](https://github.com/NVIDIA/nvidia-docker) and run one of the pre-compiled examples with
```bash
nvidia-docker run -it [...] # I tell you that cmd an other day
```

**Comment by bussmann** [comment]
Woooohooooo!

**Comment by PrometheusPi** [comment]
great job @ax3l :+1:

**Comment by ax3l** [comment]
New, spack partitioned docker recipe: https://github.com/ax3l/picongpu/tree/topic-dockerCuda8

**Comment by ax3l** [comment]
available for 0.3.0 now
